torch/_inductor/codegen/triton.py

```diff
@@ -2171,11 +2171,17 @@ def reduction_hint(node):
     def can_use_32bit_indexing(numel: sympy.Expr, buffers: Iterable[ir.Buffer]) -> bool:
         int_max = torch.iinfo(torch.int32).max
         size_hint = V.graph.sizevars.size_hint
-        if size_hint(numel) > int_max:
-            return False
+        has_hint = V.graph.sizevars.shape_env.has_hint
+
+        if not V.graph.sizevars.is_expr_static_and_true(numel <= int_max):
+            if not has_hint(numel) or size_hint(numel) > int_max:
```

Comment:
Can you please factor out this condition into its own function and use here and in the `any`.
A small comment like
```
# if we can't prove that its less than 2**32 and we have that it is an unbacked int or its current value is larger than 2**32, then we can't  use 32 bit arithmetic.
```